asllib/StaticModel.ml

```diff
@@ -2,366 +2,428 @@ open AST
+
+  let termwise f poly =
+    MonomialMap.fold
+      (fun mono factor -> add (f factor mono))
+      poly MonomialMap.empty
+
+  let mult_mono poly factor mono =
+    termwise
+      (fun f m -> single_term (Monomial.mult m mono) (Q.mul f factor))
+      poly
+
+  let neg poly = MonomialMap.map Q.neg poly
+  let mult poly1 poly2 = termwise (mult_mono poly1) poly2
+
+  let divide_by_term poly factor mono =
+    termwise
+      (fun f m -> single_term (Monomial.divide m mono) (Q.div f factor))
+      poly
+
+  let extract_constant_term poly =
+    let c = try MonomialMap.find Monomial.one poly with Not_found -> Q.zero
+    and p = MonomialMap.remove Monomial.one poly in
+    (c, p)
+
+  let is_constant poly =
+    if MonomialMap.is_empty poly then Some Q.zero
+    else if MonomialMap.cardinal poly = 1 then
+      MonomialMap.find_opt Monomial.one poly
+    else None
+
+  let to_expr poly =
+    List.fold_left
+      (fun acc (m, c) ->
+        if acc = Expr.zero then Monomial.to_expr m c
+        else
+          let e_m = Monomial.to_expr m (Q.abs c) in
+          Expr.add acc (Q.sign c, e_m))
+      Expr.zero
+      (MonomialMap.bindings poly |> List.rev)
+
+  let pp f poly =
+    let open Format in
+    if MonomialMap.is_empty poly then pp_print_string f "0"
+    else (
+      pp_open_hvbox f 2;
+      let pp_sep f () = fprintf f "@ + " in
+      PP.pp_print_seq ~pp_sep Monomial.pp_with_factor f
+        (MonomialMap.to_seq poly);
+      pp_close_box f ())
 end
 
-(** A polynomial.
+module Conjunction : sig
+  type eq = Zero | NonZero
+  type t
 
-      For example, {m X^2 - X + 4 } is represented by
-      {m X^2 \to 1, X \to -1, 1 \to 4 } *)
-type polynomial = Sum of Q.t MMap.t  (** Maps each monomial to its factor. *)
+  val empty : t
+  val is_bottom : t -> bool
+  val is_empty : t -> bool
+  val of_bool : bool -> t
+  val single_conjunct : Polynomial.t -> eq -> t
+  val conj : t -> t -> t
+  val to_expr : t -> expr option
+  val is_trivial : t -> bool option
+  val reduce : t -> t
+  val pp : Format.formatter -> t -> unit
+end = struct
+  type eq = Zero | NonZero  (** A (in)equation for a numerical value. *)
 
-module PolynomialOrdered = struct
-  type t = polynomial
+  let satisfies_eq q eq =
+    let eq_zero = Q.equal q Q.zero in
+    match eq with Zero -> eq_zero | NonZero -> not eq_zero
 
-  let compare (Sum p1) (Sum p2) = MMap.compare Q.compare p1 p2
-end
+  let pp_eq f s =
+    let s = match s with Zero -> "= 0" | NonZero -> "!= 0" in
+    Format.pp_print_string f s
 
-module PMap = Map.Make (PolynomialOrdered)
-(** Map from polynomials. *)
+  let eq_to_op = function Zero -> EQ_OP | NonZero -> NEQ
 
-(** A constraint on a numerical value. *)
-type sign =
-  | Null
-  | StrictPositive
-  | Positive
-  | Negative
-  | StrictNegative
-  | NotNull
+  module PolynomialMap = Map.Make (Polynomial)
+  (** Map from polynomials. *)
 
-(** A conjunctive logical formulae with polynomials.
+  type t = eq PolynomialMap.t option
```

Comment:
Why do you need an option here?

Reply:
I've added more to  @the comment to explain: `None` is a trivially false conjunction